### PR TITLE
PurpleVM - Datatype Conversion Operators

### LINKED ISSUE #103: Virtual Machine - Implement Datatype Conversion Operators
Datatype conversion operators are responsible for performing casts on operands. These are:
- [x] `0x86` - i32Wrapi64
- [x] `0x87` - i32TruncSignedf32
- [x] `0x88` - i32TruncUnsignedf32
- [x] `0x89` - i32TruncUnsignedf64
- [x] `0x8a` - i32TruncUnsignedf64
- [x] `0x8b` - i64ExtendSignedi32
- [x] `0x8c` - i64ExtendUnsignedi32
- [x] `0x8d` - i64TruncSignedf32
- [x] `0x8e` - i64TruncUnsignedf32
- [x] `0x8f` - i64TruncSignedf64
- [x] `0x90` - i64TruncUnsignedf64
- [x] `0x91` - f32ConvertSignedi32    
- [x] `0x92` -  f32ConvertUnsignedi32
- [x] `0x93` - f32ConvertSignedi64 
- [x] `0x94` -  f32ConvertUnsignedi64
- [x] `0x95` - f32Demotef64
- [x] `0x96` - f64ConvertSignedi32
- [x] `0x97` - f64ConvertUnsignedi32
- [x] `0x98` - f64ConvertSignedi64
- [x] `0x99` - f64ConvertUnsignedi64
- [x] `0x9a` - f64Promotef32
- [x] `0x9b` - i32Reinterpretf32
- [x] `0x9c` - i64Reinterpretf64
- [x] `0x9d` - f32Reinterpreti32
- [x] `0x9e` - f64Reinterpreti64

### PR DESCRIPTION
**Pull request by iNemesis21**
Hello,
This PR will close #103.

### CODE REVIEW

File: src/purple_vm/src/primitives/value.rs [hunk 1]
```diff
@@ -149,6 +151,22 @@ impl VmValue {
         }
     }
 
+    pub fn i32trunc_signedf32(&self) -> Result<VmValue, VmError> {
+        match *self {
+            VmValue::F32(val) => {
+                println!("before cast {:?} -> {:?}", val, val.to_be_bytes());
```
Comment (octavonce) [inline src/purple_vm/src/primitives/value.rs]: Please avoid using `println!()` statements as much as you can. While they are useful for debugging one should be very careful of leaking them into the `master` branch as they bypass any logging front-end and can result in performance bottlenecks. I suggest using one of either the `trace!()` or `debug!()` macros.

File: src/purple_vm/src/virtual_machine.rs [hunk 2]
```diff
@@ -3716,6 +3724,28 @@ fn is_type_integer(operand: VmValue) -> bool {
     !is_type_float(operand)
 }  
 
+fn perform_data_conversion(op: Instruction, operand_stack: &mut Stack<VmValue>) -> Result<(), VmError> {
+    // TODO: check the length of the operands allowed
+    let len = operand_stack.len();
+    let operand = operand_stack.pop();
+    let result:VmValue;
```
Comment (octavonce) [inline src/purple_vm/src/virtual_machine.rs]: This line is ambiguous.

File: src/purple_vm/src/primitives/value.rs [hunk 3]
```diff
@@ -149,6 +151,244 @@ impl VmValue {
...
+            _ => return Err(VmError::InvalidOperand)
+        }
+    }
+
+    // TODO: Extend for arrays
+    pub fn i32_wrapi64 (&self) -> Result<VmValue, VmError> {
+        VmValue::safe_i64_to_i32(*self)
+    }
+
+    pub fn i32trunc_f32(&self) -> Result<VmValue, VmError> {
```
Comment (octavonce) [inline src/purple_vm/src/primitives/value.rs]: The naming here seems to be flipped. The method name is `i32trunc_f32` when it should be `f32trunc_i32`. This is not intuitive and can result in serious programming errors.

File: src/purple_vm/src/primitives/value.rs [hunk 4]
```diff
@@ -149,6 +151,244 @@ impl VmValue {
...
+        match *self {
+            VmValue::F32(_) => VmValue::safe_f32_to_i32(*self),
+            _ => return Err(VmError::InvalidOperand)
+        }
+    }
+
+    pub fn i32trunc_f64(&self) -> Result<VmValue, VmError> {
```
Comment (octavonce) [inline src/purple_vm/src/primitives/value.rs]: Please maintain naming consistency.

File: src/purple_vm/src/primitives/value.rs [hunk 5]
```diff
@@ -149,6 +151,244 @@ impl VmValue {
...
+        match *self {
+            VmValue::F64(_) => VmValue::safe_f64_to_i32(*self),
+            _ => return Err(VmError::InvalidOperand)
+        }
+    }
+
+    pub fn i64extend_i32(&self) -> Result<VmValue, VmError> {
```
Comment (octavonce) [inline src/purple_vm/src/primitives/value.rs]: Please maintain naming consistency.

File: src/purple_vm/src/primitives/value.rs [hunk 6]
```diff
@@ -149,6 +151,244 @@ impl VmValue {
...
+                match val.to_i64() {
+                    Some(res) => Ok(VmValue::I64(res)),
+                    None => return Err(VmError::UnsafeCast)
+                }
+            },
+            _ => return Err(VmError::InvalidOperand)
+        }
+    }
+
+    pub fn i64trunc_f32(&self) -> Result<VmValue, VmError> {
```
Comment (octavonce) [inline src/purple_vm/src/primitives/value.rs]: Please mainting naming consistency.

File: src/purple_vm/src/primitives/value.rs [hunk 7]
```diff
@@ -149,6 +151,244 @@ impl VmValue {
...
+        match *self {
+            VmValue::F32(_) => VmValue::safe_f32_to_i64(*self),
+            _ => return Err(VmError::InvalidOperand)
+        }
+    }
+    
+    pub fn i64trunc_f64(&self) -> Result<VmValue, VmError> {
```
Comment (octavonce) [inline src/purple_vm/src/primitives/value.rs]: Please maintain naming consistency.

File: src/purple_vm/src/primitives/value.rs [hunk 8]
```diff
@@ -149,6 +151,244 @@ impl VmValue {
...
+        match *self {
+            VmValue::F64(_) => VmValue::safe_f64_to_i64(*self),
+            _ => return Err(VmError::InvalidOperand)
+        }
+    }
+
+    pub fn f32convert_i32(&self) -> Result<VmValue, VmError> {
```
Comment (octavonce) [inline src/purple_vm/src/primitives/value.rs]: Please maintain naming consistency.

File: src/purple_vm/src/primitives/value.rs [hunk 9]
```diff
@@ -149,6 +151,244 @@ impl VmValue {
...
+        match *self {
+            VmValue::I32(_) => VmValue::safe_i32_to_f32(*self),
+            _ => return Err(VmError::InvalidOperand)
+        }
+    }
+
+    pub fn f32convert_i64(&self) -> Result<VmValue, VmError> {
```
Comment (octavonce) [inline src/purple_vm/src/primitives/value.rs]: Please maintain naming consistency.

File: src/purple_vm/src/primitives/value.rs [hunk 10]
```diff
@@ -149,6 +151,244 @@ impl VmValue {
...
+        match *self {
+            VmValue::I64(_) => VmValue::safe_i64_to_f32(*self),
+            _ => return Err(VmError::InvalidOperand)
+        }
+    }
+    
+    pub fn f32demote_f64(&self) -> Result<VmValue, VmError> {
```
Comment (octavonce) [inline src/purple_vm/src/primitives/value.rs]: Please maintain naming consistency.

File: src/purple_vm/src/primitives/value.rs [hunk 11]
```diff
@@ -149,6 +151,244 @@ impl VmValue {
...
+        match *self {
+            VmValue::F64(_) => VmValue::safe_f64_to_f32(*self),
+            _ => return Err(VmError::InvalidOperand)
+        }
+    }
+
+    pub fn f64convert_i32(&self) -> Result<VmValue, VmError> {
```
Comment (octavonce) [inline src/purple_vm/src/primitives/value.rs]: Please maintain naming consistency.

File: src/purple_vm/src/primitives/value.rs [hunk 12]
```diff
@@ -149,6 +151,244 @@ impl VmValue {
...
+        match *self {
+            VmValue::I32(_) => VmValue::safe_i32_to_f64(*self),
+            _ => return Err(VmError::InvalidOperand)
+        }
+    }
+
+    pub fn f64convert_i64(&self) -> Result<VmValue, VmError> {
```
Comment (octavonce) [inline src/purple_vm/src/primitives/value.rs]: Please maintain naming consistency.

File: src/purple_vm/src/primitives/value.rs [hunk 13]
```diff
@@ -149,6 +151,244 @@ impl VmValue {
...
+        match *self {
+            VmValue::I64(_) => VmValue::safe_i64_to_f64(*self),
+            _ => return Err(VmError::InvalidOperand)
+        }
+    }
+
+    pub fn f64promote_f32(&self) -> Result<VmValue, VmError> {
```
Comment (octavonce) [inline src/purple_vm/src/primitives/value.rs]: Please maintain naming consistency.

**Comment by iNemesis21** [comment]
Hi @octavonce 
The implementation is done. All tests are added.

**Review by octavonce — APPROVED** [review]
This looks great! I rebased the branch with master and will merge it momentarily

**Pull request merged**